File: public/lexer.js
// Token type enum - must match Zig enum
const TokenType = {
    Number: 0,
    Identifier: 1,
    If: 2,
    Else: 3,
    While: 4,
    For: 5,
    Fn: 6,
    Let: 7,
    Return: 8,
    True: 9,
    False: 10,
    Plus: 11,
    Minus: 12,
    Star: 13,
    Slash: 14,
    Equal: 15,
    EqualEqual: 16,
    BangEqual: 17,
    Less: 18,
    Greater: 19,
    LessEqual: 20,
    GreaterEqual: 21,
    LeftParen: 22,
    RightParen: 23,
    LeftBrace: 24,
    RightBrace: 25,
    Semicolon: 26,
    Comma: 27,
    Eof: 28,
    Invalid: 29,
};

const TokenTypeNames = [
    'Number', 'Identifier',
    'If', 'Else', 'While', 'For', 'Fn', 'Let', 'Return', 'True', 'False',
    'Plus', 'Minus', 'Star', 'Slash', 'Equal', 'EqualEqual', 'BangEqual',
    'Less', 'Greater', 'LessEqual', 'GreaterEqual',
    'LeftParen', 'RightParen', 'LeftBrace', 'RightBrace', 'Semicolon', 'Comma',
    'Eof', 'Invalid'
];

// Token category for styling
function getTokenCategory(tokenType) {
    if (tokenType >= TokenType.If && tokenType <= TokenType.False) {
        return 'keyword';
    }
    if (tokenType === TokenType.Number) {
        return 'number';
    }
    if (tokenType === TokenType.Identifier) {
        return 'identifier';
    }
    if (tokenType >= TokenType.Plus && tokenType <= TokenType.GreaterEqual) {
        return 'operator';
    }
    if (tokenType >= TokenType.LeftParen && tokenType <= TokenType.Comma) {
        return 'delimiter';
    }
    if (tokenType === TokenType.Invalid) {
        return 'invalid';
    }
    return 'other';
}

class LexerWasm {
    constructor() {
        this.wasm = null;
        this.memory = null;
        this.exports = null;
    }

    async init() {
        try {
            const response = await fetch('lexer.wasm');
            const buffer = await response.arrayBuffer();

            const importObject = {
                env: {
                    memory: new WebAssembly.Memory({ initial: 256 }),
                },
            };

            const result = await WebAssembly.instantiate(buffer, importObject);
            this.wasm = result.instance;
            this.exports = this.wasm.exports;
            this.memory = this.exports.memory;

            return true;
        } catch (error) {
            console.error('Failed to load WASM module:', error);
            return false;
        }
    }

    tokenize(source) {
        if (!this.wasm) {
            throw new Error('WASM module not initialized');
        }

        // Clean up previous state
        this.exports.cleanup();

        // Encode source to UTF-8
        const encoder = new TextEncoder();
        const sourceBytes = encoder.encode(source);

        // Allocate memory for source
        const sourcePtr = this.exports.alloc(sourceBytes.length);
        if (!sourcePtr) {
            throw new Error('Failed to allocate memory for source');
        }

        // Copy source to WASM memory
        const memoryView = new Uint8Array(this.memory.buffer);
        memoryView.set(sourceBytes, sourcePtr);

        // Initialize lexer
        this.exports.initLexer(sourcePtr, sourceBytes.length);

        // Tokenize
        const tokenCount = this.exports.tokenizeAll();

        // Extract tokens
        const tokens = [];
        for (let i = 0; i < tokenCount; i++) {
            const type = this.exports.getTokenType(i);
            const start = this.exports.getTokenStart(i);
            const length = this.exports.getTokenLength(i);

            // Get lexeme
            const lexemePtr = this.exports.alloc(length);
            this.exports.getTokenLexeme(i, lexemePtr);

            const lexemeView = new Uint8Array(this.memory.buffer, lexemePtr, length);
            const decoder = new TextDecoder();
            const lexeme = decoder.decode(lexemeView);

            tokens.push({
                type,
                typeName: TokenTypeNames[type] || 'Unknown',
                lexeme,
                start,
                length,
            });

            this.exports.dealloc(lexemePtr, length);

            // Stop at EOF
            if (type === TokenType.Eof) break;
        }

        // Free source memory
        this.exports.dealloc(sourcePtr, sourceBytes.length);

        return tokens;
    }
}

// UI Controller
class LexerUI {
    constructor() {
        this.lexer = new LexerWasm();
        this.initialized = false;

        this.elements = {
            sourceInput: document.getElementById('sourceInput'),
            tokenizeBtn: document.getElementById('tokenizeBtn'),
            clearBtn: document.getElementById('clearBtn'),
            tokensOutput: document.getElementById('tokensOutput'),
            tokenCount: document.getElementById('tokenCount'),
            loading: document.getElementById('loading'),
        };

        this.setupEventListeners();
        this.loadDefaultExample();
    }

    async init() {
        this.showLoading(true);
        this.initialized = await this.lexer.init();
        this.showLoading(false);

        if (!this.initialized) {
            this.showError('Failed to load WASM module. Please check console for details.');
            this.elements.tokenizeBtn.disabled = true;
        }
    }

    setupEventListeners() {
        this.elements.tokenizeBtn.addEventListener('click', () => this.handleTokenize());
        this.elements.clearBtn.addEventListener('click', () => this.handleClear());

        // Allow Ctrl+Enter to tokenize
        this.elements.sourceInput.addEventListener('keydown', (e) => {
            if ((e.ctrlKey || e.metaKey) && e.key === 'Enter') {
                this.handleTokenize();
            }
        });
    }

    loadDefaultExample() {
        const example = `let x = 42;
let y = 3.14;

if (x > 10) {
    return true;
} else {
    return false;
}

fn calculate(a, b) {
    let result = a + b * 2;
    while (result >= 0) {
        result = result - 1;
    }
    return result;
}`;
        this.elements.sourceInput.value = example;
    }

    handleTokenize() {
        if (!this.initialized) {
            this.showError('WASM module not ready');
            return;
        }

        const source = this.elements.sourceInput.value;
        if (!source.trim()) {
            this.showError('Please enter some code to tokenize');
            return;
        }

        try {
            const tokens = this.lexer.tokenize(source);
            this.displayTokens(tokens);
        } catch (error) {
            console.error('Tokenization error:', error);
            this.showError('Tokenization failed: ' + error.message);
        }
    }

    handleClear() {
        this.elements.sourceInput.value = '';
        this.elements.tokensOutput.innerHTML = '<p class="placeholder">Tokens will appear here after tokenization</p>';
        this.elements.tokenCount.textContent = '0 tokens';
        this.elements.sourceInput.focus();
    }

    displayTokens(tokens) {
        const output = this.elements.tokensOutput;
        output.innerHTML = '';

        // Filter out EOF token for display
        const displayTokens = tokens.filter(t => t.type !== TokenType.Eof);

        if (displayTokens.length === 0) {
            output.innerHTML = '<p class="placeholder">No tokens found</p>';
            this.elements.tokenCount.textContent = '0 tokens';
            return;
        }

        displayTokens.forEach(token => {
            const tokenEl = document.createElement('div');
            tokenEl.className = `token ${getTokenCategory(token.type)}`;

            const typeEl = document.createElement('span');
            typeEl.className = 'token-type';
            typeEl.textContent = token.typeName;

            const lexemeEl = document.createElement('span');
            lexemeEl.className = 'token-lexeme';
            lexemeEl.textContent = token.lexeme;

            tokenEl.appendChild(typeEl);
            tokenEl.appendChild(lexemeEl);
            output.appendChild(tokenEl);
        });

        this.elements.tokenCount.textContent = `${displayTokens.length} token${displayTokens.length === 1 ? '' : 's'}`;
    }

    showError(message) {
        const output = this.elements.tokensOutput;
        output.innerHTML = `<p class="placeholder" style="color: var(--error);">${message}</p>`;
    }

    showLoading(show) {
        if (show) {
            this.elements.loading.classList.remove('hidden');
        } else {
            this.elements.loading.classList.add('hidden');
        }
    }
}

// Initialize app when DOM is loaded
document.addEventListener('DOMContentLoaded', async () => {
    const ui = new LexerUI();
    await ui.init();
});

File: src/lexer.zig
const std = @import("std");

pub const TokenType = enum {
    // Literals
    Number,
    Identifier,

    // Keywords
    If,
    Else,
    While,
    For,
    Fn,
    Let,
    Return,
    True,
    False,

    // Operators
    Plus,
    Minus,
    Star,
    Slash,
    Equal,
    EqualEqual,
    BangEqual,
    Less,
    Greater,
    LessEqual,
    GreaterEqual,

    // Delimiters
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Semicolon,
    Comma,

    // Special
    Eof,
    Invalid,
};

pub const Token = struct {
    type: TokenType,
    start: u32,
    length: u32,
};

pub const Lexer = struct {
    source: []const u8,
    start: u32,
    current: u32,

    pub fn init(source: []const u8) Lexer {
        return Lexer{
            .source = source,
            .start = 0,
            .current = 0,
        };
    }

    pub fn nextToken(self: *Lexer) Token {
        self.skipWhitespace();
        self.start = self.current;

        if (self.isAtEnd()) {
            return self.makeToken(.Eof);
        }

        const c = self.advance();

        if (isAlpha(c)) return self.identifier();
        if (isDigit(c)) return self.number();

        return switch (c) {
            '(' => self.makeToken(.LeftParen),
            ')' => self.makeToken(.RightParen),
            '{' => self.makeToken(.LeftBrace),
            '}' => self.makeToken(.RightBrace),
            ';' => self.makeToken(.Semicolon),
            ',' => self.makeToken(.Comma),
            '+' => self.makeToken(.Plus),
            '-' => self.makeToken(.Minus),
            '*' => self.makeToken(.Star),
            '/' => self.makeToken(.Slash),
            '=' => if (self.match('=')) self.makeToken(.EqualEqual) else self.makeToken(.Equal),
            '!' => if (self.match('=')) self.makeToken(.BangEqual) else self.makeToken(.Invalid),
            '<' => if (self.match('=')) self.makeToken(.LessEqual) else self.makeToken(.Less),
            '>' => if (self.match('=')) self.makeToken(.GreaterEqual) else self.makeToken(.Greater),
            else => self.makeToken(.Invalid),
        };
    }

    fn isAtEnd(self: *Lexer) bool {
        return self.current >= self.source.len;
    }

    fn advance(self: *Lexer) u8 {
        self.current += 1;
        return self.source[self.current - 1];
    }

    fn peek(self: *Lexer) u8 {
        if (self.isAtEnd()) return 0;
        return self.source[self.current];
    }

    fn match(self: *Lexer, expected: u8) bool {
        if (self.isAtEnd()) return false;
        if (self.source[self.current] != expected) return false;
        self.current += 1;
        return true;
    }

    fn skipWhitespace(self: *Lexer) void {
        while (true) {
            if (self.isAtEnd()) return;
            const c = self.peek();
            switch (c) {
                ' ', '\r', '\t', '\n' => {
                    _ = self.advance();
                },
                else => return,
            }
        }
    }

    fn makeToken(self: *Lexer, token_type: TokenType) Token {
        return Token{
            .type = token_type,
            .start = self.start,
            .length = self.current - self.start,
        };
    }

    fn number(self: *Lexer) Token {
        while (isDigit(self.peek())) {
            _ = self.advance();
        }

        // Look for decimal point
        if (self.peek() == '.' and isDigit(self.peekNext())) {
            _ = self.advance(); // consume '.'
            while (isDigit(self.peek())) {
                _ = self.advance();
            }
        }

        return self.makeToken(.Number);
    }

    fn identifier(self: *Lexer) Token {
        while (isAlpha(self.peek()) or isDigit(self.peek())) {
            _ = self.advance();
        }
        return self.makeToken(self.identifierType());
    }

    fn identifierType(self: *Lexer) TokenType {
        const lexeme = self.source[self.start..self.current];

        if (std.mem.eql(u8, lexeme, "if")) return .If;
        if (std.mem.eql(u8, lexeme, "else")) return .Else;
        if (std.mem.eql(u8, lexeme, "while")) return .While;
        if (std.mem.eql(u8, lexeme, "for")) return .For;
        if (std.mem.eql(u8, lexeme, "fn")) return .Fn;
        if (std.mem.eql(u8, lexeme, "let")) return .Let;
        if (std.mem.eql(u8, lexeme, "return")) return .Return;
        if (std.mem.eql(u8, lexeme, "true")) return .True;
        if (std.mem.eql(u8, lexeme, "false")) return .False;

        return .Identifier;
    }

    fn peekNext(self: *Lexer) u8 {
        if (self.current + 1 >= self.source.len) return 0;
        return self.source[self.current + 1];
    }
};

fn isDigit(c: u8) bool {
    return c >= '0' and c <= '9';
}

fn isAlpha(c: u8) bool {
    return (c >= 'a' and c <= 'z') or (c >= 'A' and c <= 'Z') or c == '_';
}
